### PR TITLE
Dynamic Sampling Context / Baggage continuation

### LINKED ISSUE #1476: feat(tracing) [DSC] Dynamic Sampling Context continuation
### **Describe the idea**
As a server side platform one python applications instrumented with the Sentry SDK should continue to propagate dynamic sampling context data to all outgoing requests.

The SDK should extract Dynamic Sampling Context from incoming requests, and propagating DSC to downstream SDKs.

### Requirement:
**When** a Python application receives a request which has baggage attached with sentry values,
**It shall** propagate those same value in further outgoing requests 
**so that** contextual data is preserved across the continuation of trace

_what should be sent?_

[baggage header](https://develop-b38lax7t3.sentry.dev/sdk/performance/dynamic-sampling-context/#baggage-header)

- [ ] sentry-trace_id
- [ ] sentry-public_key
- [ ] sentry-sample_rate
- [ ] sentry-release
- [ ] sentry-environment
- [ ] sentry-user_id
- [ ] sentry-user_segment
- [ ] sentry-transaction

[Trace Envelope Header](https://develop-b38lax7t3.sentry.dev/sdk/performance/dynamic-sampling-context/#envelope-header)

- [ ] trace_id (string) 
- [ ] public_key (string)
- [ ] sample_rate (string) 
- [ ] release (string) 
- [ ] environment (string) 
- [ ] user_id (string) 
- [ ] user_segment (string) 
- [ ] transaction (string) 

_where should it be sent?_

- in dynamic sample context (by baggage)
- in the envelope header
- meta data (was/is being done in JS to be check if still necessary)

_Example:_
Here is an example from the Sentry UI, where the implementation has already begun on the latest versions of the JS SDK. A request from the front end to a down stream applications includes in the request header `baggage`

image

### **Why do you think it's beneficial to most of the users**
This ensures that further downstream applications connected in a distributed trace will contain all contextual data required for dynamically sampling based on upon the contextual data from the head of the trace.

### **Possible implementation**
Implementation has already begun (at the time of this creation) on JS and Android. Refer to the SDK Developer documentation for further information

For full up to date requirements according to SDK Developer specifications please see the associated documentation.

[Dynamic Sampling Context](https://develop-b38lax7t3.sentry.dev/sdk/performance/dynamic-sampling-context/)

Original PR for full context
Add spec for Dynamic Sampling Context #613

### PR DESCRIPTION
**Pull request by sl0thentr0py**
* `Baggage` class implementing sentry/third party/mutable logic with parsing from header and serialization
* Parse incoming `baggage` header while starting transaction and store it on the transaction
* Extract `dynamic_sampling_context` fields and add to the `trace` field in the envelope header while sending the transaction
* Propagate the `baggage` header (only sentry fields / no third party as per spec)

[DSC Spec](https://develop.sentry.dev/sdk/performance/dynamic-sampling-context/)
fixes #1476

### CODE REVIEW

**Comment by sl0thentr0py** [comment]
E2E testing dump (from different runs of the same request `/api/shows/` from react)

Outgoing request from react:
```
GET /api/shows/?q=test HTTP/1.1
sentry-trace: 29fe0a09960b4c06b3d60d7dd653b5eb-96813dbb300bb498-1
baggage: sentry-environment=dev,sentry-release=0.0.1,sentry-transaction=%2Fapi%2Fshows%2F%3Fq%3Dtest,sentry-user_id=278612,sentry-user_segment=Some%20segment%20(485744)%7D,sentry-public_key=e39cc598149f46899082d1348c4abac1,sentry-trace_id=29fe0a09960b4c06b3d60d7dd653b5eb,sentry-sample_rate=1
```

[Incoming baggage header]( https://github.com/getsentry/sentry-python/blob/229e40ee0c440ce7b33513d01812c97289aed1ce/sentry_sdk/tracing.py#L234):
```
(Pdb) headers.get('baggage')
'sentry-environment=dev,sentry-release=0.0.1,sentry-transaction=%2Fapi%2Fshows%2F%3Fq%3Dtest,sentry-user_id=970676,sentry-user_segment=Some%20segment%20(763434)%7D,sentry-public_key=e39cc598149f46899082d1348c4abac1,sentry-trace_id=ad113195e0614ecca6598645db4114bd,sentry-sample_rate=1'
```

`trace` envelope header in transaction payload in `transport.capture_envelope`
```
(Pdb) envelope.headers["trace"]
{'trace_id': 'a207344557f04656bb7aeff10c725312', 'public_key': 'e39cc598149f46899082d1348c4abac1', 'sample_rate': '1', 'release': '0.0.1', 'environment': 'dev', 'transaction': '/api/shows/?q=tes', 'user_id': '970676', 'user_segment': 'Some segment (763434)}'}
```

File: sentry_sdk/tracing_utils.py [hunk 1]
```diff
@@ -413,6 +418,85 @@ def has_custom_measurements_enabled():
...
+    def freeze(self):
+        # type: () -> None
+        self.mutable = False
+
+    def dynamic_sampling_context(self):
+        # type: () -> Dict[str, str]
+        header = {}
+
+        for key in Baggage.DSC_KEYS:
+            if self.sentry_items.get(key):
```
Comment (antonpirker) [inline sentry_sdk/tracing_utils.py:480]: sentry_items is accessed twice. would saving into local variable and then do the if against that var and saving that var in `header` make things more performant?

Comment (sl0thentr0py) [reply sentry_sdk/tracing_utils.py:480]: it's a variable on the object anyway, don't think that's necessary

File: sentry_sdk/tracing_utils.py [hunk 2]
```diff
@@ -413,6 +418,85 @@ def has_custom_measurements_enabled():
...
+        if header:
+            for item in header.split(","):
+                item = item.strip()
+                key, val = item.split("=")
+                if Baggage.SENTRY_PREFIX_REGEX.match(key):
+                    baggage_key = unquote(key.split("-")[1])
+                    sentry_items[baggage_key] = unquote(val)
+                    mutable = False
+                else:
+                    third_party_items += val
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply sentry_sdk/tracing_utils.py]: thx, fixed and I'll write tests for this too
```
>>> from sentry_sdk.tracing_utils import Baggage
>>> b = Baggage.from_incoming_header("other-vendor-value-1=foo;bar;baz, sentry-trace_id=771a43a4192642f0b136d5159a501700, sentry-public_key=49d0f7386ad645858ae85020e393bef3, sentry-sample_rate=0.01337, sentry-user_id=Am%C3%A9lie, other-vendor-value-2=foo;bar;")
>>> b.serialize()
'sentry-trace_id=771a43a4192642f0b136d5159a501700,sentry-public_key=49d0f7386ad645858ae85020e393bef3,sentry-sample_rate=0.01337,sentry-user_id=Am%C3%A9lie'
>>> b.serialize(include_third_party=True)
'sentry-trace_id=771a43a4192642f0b136d5159a501700,sentry-public_key=49d0f7386ad645858ae85020e393bef3,sentry-sample_rate=0.01337,sentry-user_id=Am%C3%A9lie,other-vendor-value-1=foo;bar;baz,other-vendor-value-2=foo;bar;'
```

File: sentry_sdk/tracing_utils.py (hunk 1 above)
Comment (sl0thentr0py) [reply sentry_sdk/tracing_utils.py:480]: oh sorry you meant the hash access, yea will do

**Review by vladanpaunovic — APPROVED** [review]
LGTM

**Pull request merged**